ophyd/areadetector/base.py

```diff
@@ -148,3 +150,117 @@ def default_match(attr, signal, doc):
 
             if match:
                 match_fcn(attr=attr, signal=getattr(self, attr), doc=doc)
+
+    def stage(self, *args, **kwargs):
+        if not self.validate_asyn_ports():
+            missing_plugins = self.missing_plugins()
+            raise RuntimeError('The asyn ports {!r} are used by plugins '
+                               'that ophyd is aware of but the source plugin '
+                               'is not.  Please reconfigure your device to '
+                               'include the source plugin or reconfigure '
+                               'to not use these ports.'
+                               ''.format(missing_plugins))
+        return super().stage(*args, **kwargs)
+
+    def get_plugin_by_asyn_port(self, port_name):
+        '''Get the plugin which has the given asyn port name
+
+        Parameters
+        ----------
+        port_name : str
+            The port name to search for
+
+        Returns
+        -------
+        ret : ADBase or None
+            Either the requested plugin or None if not found
+
+        '''
+        try:
+            name = self.port_name.get()
+            if name == port_name:
+                return self
+        except AttributeError:
+            pass
+
+        signames = self._sub_devices
+
+        for nm in signames:
+            cpt = getattr(self, nm)
+            if hasattr(cpt, 'get_plugin_by_asyn_port'):
+                sig = cpt.get_plugin_by_asyn_port(port_name)
+                if sig is not None:
+                    return sig
+        return None
+
+    def get_asyn_port_dictionary(self):
+        '''Return port name : component map
+
+        Returns
+        -------
+        port_map : dict
+            Mapping between port_name and ADBase objects
+        '''
+        # uniqueness of port names enforced at IOC layer
+        ret = {}
+        try:
+            ret.update({self.port_name.get(): self})
+        except AttributeError:
+            pass
+
+        signames = self._sub_devices
+
+        for nm in signames:
+            sig = getattr(self, nm)
+            if hasattr(sig, 'get_asyn_port_dictionary'):
+                ret.update(sig.get_asyn_port_dictionary())
+
+        return ret
+
+    def get_asyn_digraph(self):
+        '''Get the directed graph of the ASYN ports
+
+        Returns
+        -------
+        G : networkx.DiGraph
+            Directed graph of pipelines
+
+        port_map : dict
+            Mapping between port_name and ADBase objects
+        '''
+        port_map = self.get_asyn_port_dictionary()
+        G = nx.DiGraph()
+        for out_port, cpt in port_map.items():
+            try:
+                in_port = cpt.nd_array_port.get()
+                G.add_edge(in_port, out_port)
```

Comment:
`G.add_edge(...)` should go in an `else` clause to be clear what the `except AttributeError` is trying to catch